util/checksum/checksum_test.go

```diff
@@ -71,3 +72,537 @@ func (s *testChecksumSuite) TestChecksumReadAt(c *check.C) {
 	assertReadAt(5, nil, 10, "5678901234")
 	assertReadAt(int64(n1+n2)-5, io.EOF, 5, "56789\x00\x00\x00\x00\x00")
 }
+
+type mockWriter struct {
+	err    error
+	w      io.WriteCloser
+	offset int
+	f      func(b []byte, offset int) []byte
+}
+
+func newMockWriter(w io.WriteCloser, f func(b []byte, offset int) []byte) *mockWriter {
+	return &mockWriter{w: w, f: f}
+}
+
+func (w *mockWriter) Write(p []byte) (n int, err error) {
+	// always write successfully.
+	n = len(p)
+	if w.f != nil {
+		p = w.f(p, w.offset)
+	}
+	nn, err := w.w.Write(p)
+	if err != nil {
+		return n, err
+	}
+	w.offset += nn
+	return n, err
+}
+
+func (w *mockWriter) Close() (err error) {
+	if w.err != nil {
+		return w.err
+	}
+	return w.w.Close()
+}
+
+func (s *testChecksumSuite) TestTiCase3644(c *check.C) {
+	s.testTiCase3644(c, false)
+	s.testTiCase3644(c, true)
+}
+
+func (s *testChecksumSuite) testTiCase3644(c *check.C, encrypt bool) {
```

Comment:
It'll be better to add some comment for `testTiCase36xx`, otherwise, it's hard to understand what the test does individually